Modal truncation

meretp opened the pull request:
This PR adds a new reductor who uses modal truncation to reduce the given model. A small example on how to use the reductor is given in the demo for the heat equation.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,156 @@
+            -`'eigs'`: use scipy.linalg.eig algorithm
+            -`'samdp'`: find dominant poles using samdp algorithm
+        projection
+            Projection method:
+            - `'orth'`: projection matrices are orthogonalized with
```
There should be an empty line before the start of a list.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,156 @@
+            - `'biorth'`: projection matrices are biorthogolized with
+              respect to the E product
+        symmetric
+            True if Operator A is symmetric, False if not
```
`E` should also be symmetric positive definite, right?

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,156 @@
+        Parameters
+        ----------
+        r
+            Order of the reduced model
```
This should be treated as a sentence, and so should end with a dot. Similar below.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,156 @@
+        rom
+            Reduced-order model.
+        """
+        assert min(self.fom.B.source.dim, self.fom.C.range.dim) <= r < self.fom.order
```
Why is this necessary?

meretp replied:
I think its not explicitly necessary but more an assumption to get good results. But that should not be an assertion since it is not necessary for the algorithm to work, right?

pmli replied:
Yes. Also, it might be possible to get good results even when the number of inputs and outputs is large, it's difficult to say in advance.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,156 @@
+        self.W = self.fom.C.as_source_array().empty(reserve=r)
+
+        if decomposition == 'eigs':
+            if self.fom.A.sparse:
+                A = self.fom.A.as_source_array().to_numpy()
+            else:
+                A = self.fom.A.matrix
```
It is more convenient to use `to_matrix` from `pymor.algorithms.to_matrix` to convert an `Operator` to a matrix.

Also, large `Operators` should not be converted silently to dense matrices, or at all. See how [`LTIModel.poles`](https://github.com/pymor/pymor/blob/2f7955ac6df214fbe1cb5d7b130c5d00e6d9bd9a/src/pymor/models/iosys.py#L621-L626) does it.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,156 @@
+            which = method_options['which']
+            if symmetric:
+                poles, ev_r = eig(A, right=True)
+                if which == 'SM':
+                    idx = np.argsort(np.abs(poles))
+                elif which == 'LR':
+                    idx = np.argsort(-poles.real)
```
These options (`'SM'` and `'LR'`) should be documented.

Also, there should be options similar to `samdp`, which sorts by `residue_norms / poles.real` or similar.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,156 @@
+                A = self.fom.A.matrix
+            which = method_options['which']
+            if symmetric:
+                poles, ev_r = eig(A, right=True)
```
`E` should be used.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,156 @@
+                    idx = np.argsort(-poles.real)
+                rev = ev_r[:, idx]
+                poles = poles[:r]
+                rev = NumpyVectorArray(rev[:, :r].T,
+                                       self.fom.B.as_range_array().space)
```
It's nicer to write something like `self.fom.A.source.from_numpy(rev[:, :r].T)`.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,156 @@
+                                       self.fom.B.as_range_array().space)
+                lev = NumpyVectorArray(lev[:, :r].T,
+                                       self.fom.B.as_range_array().space)
+
+        elif decomposition == 'samdp':
+
```
Empty lines here can be removed.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,156 @@
+                                         self.fom.C.as_source_array(),
+                                         r, **method_options)
+        if np.iscomplexobj(poles):
+
```
This empty line can be removed. Similarly the one before `else:`.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,156 @@
+                                       for op in ['A', 'B', 'C', 'D', 'E']})
+        else:
+            fom_mu = self.fom
+        self._pg_reductor = LTIPGReductor(fom_mu, self.W,
+                                          self.V, projection == 'biorth')
```
This can be in the same line.

pmli commented on src/pymordemos/heat.py:
```diff
@@ -110,6 +110,26 @@
+    # Modal Truncation
+    r = 5
+    modal_reductor = MTReductor(lti)
+    rom_mt = modal_reductor.reduce(r, decomposition = 'samdp')
```
There is a `flake8` warning here (spaces around `=` in keyword argument).

pmli commented on src/pymordemos/heat.py:
```diff
@@ -110,6 +110,26 @@
+    rom_mt.mag_plot(w, ax=ax, linestyle='dashed')
+    ax.set_title('Magnitude plot of the full and MT reduced model')
+    plt.show()
+    
```
It seems there are extra spaces here (`flake8` warning).

pmli commented on src/pymor/basic.py:
```diff
@@ -65,6 +65,7 @@
 
 from pymor.reductors.basic import StationaryRBReductor, InstationaryRBReductor, LTIPGReductor, SOLTIPGReductor
 from pymor.reductors.bt import BTReductor, LQGBTReductor, BRBTReductor
+from pymor.reductors.mt import MTReductor
```
This import should also be sorted.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,169 @@
+        if not method_options:
+            method_options = {}
+
+        self.V = self.fom.B.as_range_array().empty(reserve=r)
+        self.W = self.fom.C.as_source_array().empty(reserve=r)
```
It's good to avoid using `as_range_array` or `as_source_array` if possible. Here, you can directly access the necessary subspace, e.g., with `self.fom.A.source.empty(reserve=r)`. (Also, it is guaranteed that `A.source == A.range` for `LTIModels`.)

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,169 @@
+        self.W = self.fom.C.as_source_array().empty(reserve=r)
+
+        if decomposition == 'eig':
+            if self.fom.A.sparse:
+                self.logger.warning('Converting operator A to a NumPy array.')
+            A = to_matrix(self.fom.A, format='dense')
+            if not isinstance(self.fom.E, IdentityOperator):
+                if self.fom.E.sparse:
+                    self.logger.warning('Converting operator E to a NumPy array.')
+            E = None if isinstance(self.fom.E, IdentityOperator) else to_matrix(self.fom.E, format='dense')
```
To be consistent with `LTIModel.poles`, it would be good to use `pymor.models.iosys.sparse_min_size` in a similar way. You can do `from pymor.models.iosys import sparse_min_size` at the top and then just copy-paste the code from `LTIModel.poles`.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,169 @@
+            Reduced-order model.
+        """
+        assert 0 < r < self.fom.order
+        assert projection in ('orth', 'biorth')
+        assert decomposition in ('eig', 'samdp')
```
It would be good to add a similar `assert` statement for `which`.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,169 @@
+            else:
+                absres = np.empty(len(poles))
+                for i in range(len(poles)):
+                    b_norm = spla.norm(lev[i].inner(self.fom.B.as_range_array()), ord=2)
```
It looks mathematically correct, but it should be possible to avoid using `as_range_array`. Try using `self.fom.B.apply_adjoint` and the `norm` method of `VectorArrays`.

Similar for `c_norm` below.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,169 @@
+            Optional dict with more options for the samdp algorithm.
+        which
+            A string specifying which `k` eigenvalues and eigenvectors to
+            compute when using the eig decomposition, default is "LR":
```
As far as I understand, it would be good to use `'NR'` as the default because it is motivated by an Hinf error bound.

meretp replied:
Okay, I choosed 'LR' because It was stated as the "classical approach" for modal truncation in the literature.

pmli replied:
Well, since you added the `'NR'` version and it's the default for `samdp`, it makes sense to make it the default for `MTReductor` :)

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,169 @@
+                    idx = np.argsort(-absres / np.abs(poles))
+                elif which == 'NM':
+                    idx = np.argsort(-absres)
+            poles = poles[idx]
+            rev = rev[idx]
+            lev = lev[idx]
+            poles = poles[:r]
+            rev = rev[:r]
+            lev = lev[:r]
```
I'm afraid weird things might happen if `poles[:r]` is not closed under conjugation (i.e., if there is a complex eigenvalue, but not also it's complex conjugate). Maybe you can try testing it with an `LTIModel` that has `-1`, `-2 + 1j`, `-2 - 1j`, `-3` as poles and residues all equal, and see what happens when you try reducing it to order 2. I'm actually not sure which poles to pick in such a case. At least in some cases, the ROM order should be exactly `r` (e.g., if modal truncation would be used inside SOR-IRKA instead of IRKA), so completing `poles[:r]` under conjugation would not be good...

meretp replied:
Yes, this is actually a problem I already thought about when comparing eig and samdp decomposition. The samdp algorithm returns the first r dominant poles closed under conjunction such that the rom would be of higher order than r for complex conjugated pairs of poles. So but as you said it is more important that the ROM order is exactly r? 
Then I will do some research which poles to pick in the case described above.

pmli replied:
Ah, right. Then it might be strange for the eig and samdp versions to behave differently. Although, SOR-IRKA would only use the dense version.
Maybe @lbalicki has some ideas.

lbalicki replied:
In my opinion with either option the order of the ROM should be `r`. I'm not aware of any experiments/results which indicate if it is OK to use only one eigenvector of a complex conjugate pair when projecting the matrices. I think we should just test some examples where this case occurs (e.g. the example @pmli mentioned) and then see if we get good results if we force the order of the ROM to be exactly `r`.

pmli replied:
I was curious about what sssMOR does. It looks like it just [warns](https://github.com/MORLab/sssMOR/blob/2f90150e138d0ec9e9c8cfc616c9aac250fcac7b/src/MOR/classic/modalMor.m#L244-L262) when the complex poles do not come in pairs and returns a complex ROM... I guess there could be situations when it's not possible to do anything else (e.g., if the FOM only has complex poles and `r` is odd), but maybe it's possible to do something otherwise (it reminds me of the knapsack problem...).

meretp replied:
So I tested the example above and compared the relative h2 error.
At first I reduced the model to order 2 with the current implementation that returns a real reduced model:
`rel. error for samdp 0.344 ,eig 0.430`
second I reduced the model to order 2 but now taking the complex eigenvector of the chosen pole from the pair: 
`rel. error for samdp 0.563, eig 0.615`
third I reduced the model to order 3:
` rel. error for samdp 0.144, eig 0.203`

From this first impression my suggestion is to take the current implementation which results in a real reduced model and would take just one pole of the complex pair but then warn the user such that the user needs to decide if he increases the reduced order by one to avoid the splitting of the pair. But I would try to test a bit more with larger problems. Do you have more suggestions which model could be good for testing?

lbalicki replied:
> But I would try to test a bit more with larger problems. Do you have more suggestions which model could be good for testing?

I would suggest testing with the [Slycot benchmarks](http://slicot.org/20-site/126-benchmark-examples-for-model-reduction). You can get the `LTIModel` from the .mat files with the `LTIModel.from_mat_file` function. Many of the models there have complex poles, e.g. the first few dominant poles (computed with samdp and which='LR') of the FOM model are complex conjugate pairs. I would suggest doing some tests with that model and some low orders for the ROM such as 3, 4, 5.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,210 @@
+        if np.iscomplexobj(poles):
+            real_index = np.where(np.isreal(poles))[0]
+            complex_index = np.where(poles.imag > 0)[0]
+            if len(np.where(poles[:r].imag)[0]) % 2 == 1:
```
I guess this assumes that the conjugate pairs in `poles` are always next to each other. I'm not sure this is true when there are two pairs with the same real parts, e.g.,

```python
>>> import numpy as np
>>> a = np.array([1j, -1j, 2j, -2j])
>>> a
array([ 0.+1.j, -0.-1.j,  0.+2.j, -0.-2.j])
>>> a.sort()
>>> a
array([-0.-2.j, -0.-1.j,  0.+1.j,  0.+2.j])
```

It might be good to test the algorithm on the FOM example from the [SLICOT benchmark collection](http://slicot.org/20-site/126-benchmark-examples-for-model-reduction).

meretp replied:
Not really. Or respectively the construction of the projection matrices takes care of it. So when constructing the projection matrices V and W the real and imaginary part of the eigenvector from the first complex pole that occurs are taken. Such that for the example above the reduced model would contain the complex pair (2j, -2j) even though the first two poles after sorting are -2j and -1j. On the FOM example it works as expected and gives a warning when reducing to order 3 for example.

pmli replied:
I'm thinking more about the example where `r` is 2. I guess then this `if` branch thinks that `-2j` and `-1j` form a conjugate pair.

meretp replied:
True, but the construction later on takes the real and imaginary part of the eigenvector of `-2j`  such that `-1j` would not be a pole of the reduced model.

pmli replied:
Hm, how? If `r` is 2, then the number of complex numbers is even, so the execution goes to the `else` part where `complex_index = np.where(poles.imag > 0)[0]`, right?

meretp replied:
Right, but yeah now I see. The selection of the complex_index needs to be the same as in the else part if the pair is split. So not only checking if `poles.imag >0` but also `poles.imag <0` and taking the index list which is longer. With that changed it should work, right?

pmli replied:
I'm not sure what you mean. Both `poles.imag > 0` and `poles.imag < 0` should be of the same length. I guess you mean `poles[:r].imag`?
Either way, the condition in this `if` branch (`len(np.where(poles[:r].imag)[0]) % 2 == 1`) should not assume that `poles[:r]` are the "right" poles. Maybe you can look at `poles[poles.imag >= 0]` and count the first `r` poles while double-counting the complex poles, e.g., with something like `np.cumsum((poles[poles.imag >= 0].imag > 0).astype(int) + 1)`.

meretp replied:
Right. 
In my opinion that would make this condition more complicated than necessary since one would still need to look at `poles.imag[:r] > 0` and `poles.imag[:r] < 0` when constructing the projection matrices. So my suggestion is to use the same construction as in lines 164-178 in the else branch of this if statement.

                    real_index = np.where(np.isreal(poles))[0]
                    complex_index = np.where(poles.imag > 0)[0] \
                        if np.sum(np.where(poles.imag > 0)) > np.sum(np.where(poles.imag < 0)) \
                        else np.where(poles.imag < 0)[0]

                    self.V.append(rev[real_index].real)
                    self.V.append(rev[complex_index].real)
                    self.V.append(rev[complex_index].imag)

                    self.W.append(lev[real_index].real)
                    self.W.append(lev[complex_index].real)
                    self.W.append(lev[complex_index].imag)

                    self.V = self.V[:r]
                    self.W = self.W[:r]

With that the if condition would still assume that `poles[:r]` are the "right" poles but I don't see why this is a problem as long as the poles or more the eigenvectors taken for the projection are the right ones. 

Another idea I had was to use stable sorting algorithms to make sure that the poles would come in pairs but I guess that would make things less efficient.

pmli replied:
I'm not sure the above approach chooses the "right" eigenvectors. As far as I see, for `poles = [-2j, -1j, 1j, 2j]` and `r = 2`, it would only take the real parts of the eigenvectors corresponding to `-2j` and `-1j`, and it should take the real and imaginary part of one of those. For `poles = [-3j, -2j, -1j, 1j, 2j, 3j]` and `r = 3`, it's similar, but it should take the real and imaginary part from one complex pair and something from another pair.

A stable sorting algorithm would work if `poles` starts out with complex pairs next to each other. That might be true for `eig`, but I'm not sure it is for `samdp`. (Efficiency matters less for reduced quantities.) You could use something like

```python
idx = sorted(
    range(len(poles)),
    key=lambda i: (-dominance[i],
                   -poles[i].real,
                   abs(poles[i].imag),
                   0 if poles[i].imag >= 0 else 1))
poles = poles[idx]
```

assuming `dominance[i] == dominance[j]` when `poles[i]` and `poles[j]` form a complex conjugate pair (or this could be ensured in the code).

meretp replied:
Ah, you're absolutely right. I was a fool, sorry for that. 
So one could also force the sorting in the `samdp` algorithm to use a stable algorithm and do the same for the sorting after the use of the `eig` algorithm. Or is there a reason why the sorting in samdp is not stable @lbalicki? Then the complex pairs would be next to each other, might be the easiest way.

pmli replied:
Well, you'll need this special sorting whether `samdp` does it or not since it's needed for the `eig` approach (assuming you go for the sorting approach). It might be good for `samdp` to do this sorting, but I guess it's not strictly necessary.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,237 @@
+        self._pg_reductor = None
+
+    def reduce(self, r=None, decomposition='samdp', projection='orth',
+               symmetric=False, which="NR", method_options=None, take_complex=False):
```
```suggestion
               symmetric=False, which='NR', method_options=None, take_complex=False):
```

It is more "pyMOR-ic" to use single quotes for strings, unless the string itself contains single quotes.

A similar comment applies to a few places below.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,210 @@
+            Algortihm to use for the decomposition:
+
+            - `'eig'`: use scipy.linalg.eig algorithm
+            - `'samdp'`: find dominant poles using samdp algorithm
```
It would be good to add the link for documentation.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,210 @@
+            - `'biorth'`: projection matrices are biorthogolized with
+              respect to the E product
+        symmetric
+            `True` if Operator A is symmetric and E is symmetric, positive
```
I rarely, if ever, see a comma here.

Also, for boolean arguments, it's common in pyMOR to document them with "If `True`, ..." (e.g., see `gram_schmidt`). Here it could be "If `True`, assume A is symmetric and E is symmetric positive definite".

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,237 @@
+            - `'NM'`: select eigenvalues with largest norm(residual)
+        method_options
+            Optional dict with more options for the samdp algorithm.
+        take_complex
```
To me, `take_complex` doesn't sound very descriptive. I would suggest `allow_complex_rom`.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,237 @@
+        assert which in ('LR', 'SM', 'NR', 'NS', 'NM')
+        assert method_options is None or isinstance(method_options, dict)
+        if not method_options:
+            method_options = {'which':'LR'}
```
flake8 issue

```suggestion
            method_options = {'which': 'LR'}
```

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,237 @@
+            E = None if isinstance(fom.E, IdentityOperator) else to_matrix(fom.E, format='dense')
+
+            if symmetric:
+                poles, ev_r = spla.eig(A, E, right=True)
```
`right=True` is not necessary (it is the default).

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,237 @@
+                poles, ev_l, ev_r = spla.eig(A, E, left=True, right=True)
+                rev = fom.A.source.from_numpy(ev_r.T)
+                lev = fom.A.source.from_numpy(ev_l.T)
+            if which == 'SM':
+                idx = np.argsort(np.abs(poles))
+                dominance = np.abs(poles)
+            elif which == 'LR':
+                idx = np.argsort(-poles.real)
+                dominance = np.ones(len(poles))
+            else:
+                absres = np.empty(len(poles))
+                for i in range(len(poles)):
+                    lev[i].scal(1 / lev[i].inner(fom.E.apply(rev[i]))[0][0])
+                    b_norm = fom.B.apply_adjoint(lev[i]).norm()
+                    c_norm = fom.C.apply(rev[i]).norm()
+                    absres[i] = b_norm @ c_norm
+                if which == 'NR':
+                    dominance = -(absres / np.abs(np.real(poles)))
+                    idx = np.argsort(dominance)
+
+                elif which == 'NS':
+                    dominance = -(absres / np.abs(poles))
+                    idx = np.argsort(dominance)
+
+                elif which == 'NM':
+                    idx = np.argsort(-absres)
+                    dominance = -np.array(absres)
```
This could be simplified a bit by removing `idx = ...` from every `if` and `elif` and just having `idx = np.argsort(dominance)` at the end.

Also, the empty lines above `elif` doesn't look like pyMOR style to me.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,237 @@
+                absres = np.empty(len(poles))
+                for i in range(len(poles)):
+                    lev[i].scal(1 / lev[i].inner(fom.E.apply(rev[i]))[0][0])
+                    b_norm = fom.B.apply_adjoint(lev[i]).norm()
+                    c_norm = fom.C.apply(rev[i]).norm()
+                    absres[i] = b_norm @ c_norm
```
Using `@` for 1x1 matrices seems a bit strange.
```suggestion
                    b_norm = fom.B.apply_adjoint(lev[i]).norm()[0]
                    c_norm = fom.C.apply(rev[i]).norm()[0]
                    absres[i] = b_norm * c_norm
```

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,237 @@
+                                         fom.B.as_range_array(),
+                                         fom.C.as_source_array(),
+                                         r, **method_options)
+            absres = [spla.norm(res[i], ord=2) for i in range(len(poles))]
```
Using the `axis` argument of the `norm` should do the same.
```suggestion
            absres = spla.norm(res, axis=(1, 2), ord=2)
```

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,237 @@
+            rev = rev[idx]
+            lev = lev[idx]
+
+            real_index = np.where(np.isreal(poles))[0]
+            complex_index = np.where(poles.imag > 0)[0]
```
It seems this can be removed.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,237 @@
+                    self.W = lev[:r]
+                else:
+                    self.logger.info("Only real part of complex conjugated pair taken.")
+                    poles = poles[:r]
+                    lev = lev[:r]
+                    rev = rev[:r]
```
This is basically repeated in the `if` above and `else` below. It could be moved to before `if len(np.where(poles[:r].imag)[0]) % 2 == 1:`.

pmli reviewed (changes requested):
I just have comments on the code style. The algorithm itself seems ok.

It would be good to also run `MTReductor` in the string and parametric_string demos.

meretp commented:
> I just have comments on the code style. The algorithm itself seems ok.

Thanks! I changed the code accordingly.
 
> It would be good to also run `MTReductor` in the string and parametric_string demos.

Working on that, I encountered that the poles from samdp are not necessarily real if the imaginary part is too small to build a complex conjugated pair. Therefore I changed the evaluation of `real_index` and `complex_index` (lines 190-193) using the default tolerance for the imaginary part from samdp.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,238 @@
+                fom.E, format='dense')
+
+            if symmetric:
+                poles, ev_r = spla.eig(A, E)
```
It would be better to used `eigh` here.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,238 @@
+                lev = lev[idx]
+                dominance = dominance[idx]
+
+                poles = poles[:2*r]
+                lev = lev[:2*r]
+                rev = rev[:2*r]
+                dominance = dominance[:2*r]
```
I guess, if `poles` is a real object, the ROM would end up being of order `2*r`.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,238 @@
+                    self.V = rev
+                    self.W = lev
+                else:
+                    self.logger.info(
+                        'Only real part of complex conjugated pair taken.')
+                    real_index = np.where(
+                        np.abs(poles.imag) / np.abs(poles) < 1e-6)[0]
+                    complex_index = np.where(
+                        (np.abs(poles.imag) / np.abs(poles) >= 1e-6) & (poles.imag > 0))[0]
```
All of these three could be in their own line.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,238 @@
+                    self.V = self.V[:r]
+                    self.W = self.W[:r]
+            else:
+                real_index = np.where(
+                    np.abs(poles.imag) / np.abs(poles) < 1e-6)[0]
+                complex_index = np.where(
+                    (np.abs(poles.imag) / np.abs(poles) >= 1e-6) & (poles.imag > 0))[0]
+
+                self.V.append(rev[real_index].real)
+                self.V.append(rev[complex_index].imag)
+                self.V.append(rev[complex_index].real)
+
+                self.W.append(lev[real_index].real)
+                self.W.append(lev[complex_index].imag)
+                self.W.append(lev[complex_index].real)
```
It would be good to reduce code repetition.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,238 @@
+                    dominance = -(absres / np.abs(poles))
+                elif which == 'NM':
+                    dominance = -np.array(absres)
+                idx = np.argsort(dominance)
```
It might be good to have sorting in only one place. Also, this sorting is not used when `which` is `'SM'` or `'LR'`.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,203 @@
+        if decomposition == 'eig':
+            if fom.order >= sparse_min_size():
+                if not isinstance(fom.A, NumpyMatrixOperator) or fom.A.sparse:
+                    self.logger.warning(
+                        'Converting operator A to a NumPy array.')
```
This could be in one line.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,203 @@
+                        'Converting operator A to a NumPy array.')
+                if not isinstance(fom.E, IdentityOperator):
+                    if not isinstance(fom.E, NumpyMatrixOperator) or fom.E.sparse:
+                        self.logger.warning(
+                            'Converting operator E to a NumPy array.')
```
This could be in one line.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,203 @@
+                        self.logger.warning(
+                            'Converting operator E to a NumPy array.')
+            A = to_matrix(fom.A, format='dense')
+            E = None if isinstance(fom.E, IdentityOperator) else to_matrix(
+                fom.E, format='dense')
```
This could be in one line.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,203 @@
+        idx = sorted(range(len(poles)),
+                     key=lambda i: (dominance[i], -poles[i].real,
+                                    abs(poles[i].imag), 0 if poles[i].imag >= 0 else 1))
+        poles = poles[idx]
+        rev = rev[idx]
+        lev = lev[idx]
+        poles = poles[:r]
+        lev = lev[:r]
+        rev = rev[:r]
```
This could be shortened to:

```python
idx = idx[:r]
poles = poles[idx]
rev = rev[idx]
lev = lev[idx]
```

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,203 @@
+        complex_index = np.where((np.abs(poles.imag) / np.abs(poles) >= 1e-6) & (poles.imag > 0))[0]
+
+        if complex_index.size > 0 and complex_index[-1] == r-1:
+            self.logger.warning(
+                'Chosen order r will split complex conjugated pair of poles.')
```
This could be in one line.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,203 @@
+                'Chosen order r will split complex conjugated pair of poles.')
+            if allow_complex_rom:
+                self.logger.info('Reduced model will be complex.')
+                self.V.append(rev[r-1])
+                self.W.append(lev[r-1])
```
Using `-1` to index the last element is a bit nicer.
Similar below, in the `else` branch.

pmli reviewed (approved):
This looks very good to me. I would be for merging after the minor edits.

pmli commented:
@renefritze It seems there is an issue with the markdown link checker. It looks like it wants to access the master branch.

@lbalicki If you want, you can go ahead with your review.

renefritze commented:
> @renefritze It seems there is an issue with the markdown link checker. It looks like it wants to access the master branch.
> 

Thanks for the heads up. https://github.com/pymor/pymor/pull/1244 should fix that

lbalicki commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,197 @@
+        assert which in ('LR', 'SM', 'NR', 'NS', 'NM')
+        assert method_options is None or isinstance(method_options, dict)
+        if not method_options:
+            method_options = {'which': 'LR'}
```
I think the default `which` argument for `samdp` should depend on the `which` argument of the `MTReductor.reduce` method. E.g. if `which == 'NS'` then the default option should be `method_options = {'which': 'LS'}` instead of `method_options = {'which': 'LR'}` in order to ensure that `samdp` returns exactly the poles that we are looking for. Similarly `method_options` should be adjusted if  `which == 'NM'` or `which == 'NR'`.

For the 'LR' and 'SM' cases I would use `method_options = {'which': 'LM'}`.

meretp replied:
Thanks for your review. 
As stated in the docstring my intention was that the which argument is only used when the eig decomposition is used otherwise the sorting for samdp should be configured with the method_options dict. But you may be right that it seems more intuitive to use the which argument for both decomposition algorithms.

pmli replied:
I had been thinking about it. It would indeed be nice if `which` can be used for both algorithms, especially once support for `eigs` is added, but it is a bit unfortunate that the option strings do not match. @lbalicki What do you think about changing the `which` options in `samdp`?

Then, if `decomposition == 'samdp'`, I would propose raising a `ValueError` if `which in ('LR', 'SM')` (and similarly for the potential future `eigs` support).

lbalicki replied:
I did the renaming in  #1247, this way the `which` argument of `MTReductor.reduce` can be passed to `samdp`. I think it will be more straightforward for a user if `which` is used for all the methods that compute the poles. I guess further arguments for `samdp` can still be passed through the `method_options`.

lbalicki reviewed (commented):
Looks very good! I just added one comment about the `which` argument of `samdp`.

sdrave commented:
@meretp, you should not forget to add yourself to `AUTHORS.md` before this gets merged ...

pmli commented on AUTHORS.md:
```diff
@@ -9,6 +9,11 @@
 # Contributors
 
+## pyMOR 2021.1
+* Meret Behrens, mbehrens@mpi-magdeburg.mpg.de
+
```
It would be nice to keep the consistency regarding empty lines.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,200 @@
+                elif which == 'NM':
+                    dominance = -np.array(absres)
+        elif decomposition == 'samdp':
+            assert which in ('NR', 'NS', 'NM')
```
It is good to have such checks closer to the beginning of the function.

pmli commented:
This PR probably needs to be rebased on the current main branch for CI to pass.

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,201 @@
+class MTReductor(BasicObject):
+    """Modal Truncation reductor.
+
+    See Section 9.2 in [A05]_.
```
```suggestion
    See Section 9.2 in :cite:`A05`.
```

pmli commented on src/pymor/reductors/mt.py:
```diff
@@ -0,0 +1,201 @@
+                                         fom.C.as_source_array(),
+                                         r, **method_options)
+            absres = spla.norm(res, axis=(1, 2), ord=2)
+            if method_options['which'] == 'LR':
+                dominance = -(absres / np.abs(np.real(poles)))
+            elif method_options['which'] == 'LS':
+                dominance = -(absres / np.abs(poles))
+            elif method_options['which'] == 'LM':
```
```suggestion
            if method_options['which'] == 'NR':
                dominance = -(absres / np.abs(np.real(poles)))
            elif method_options['which'] == 'NS':
                dominance = -(absres / np.abs(poles))
            elif method_options['which'] == 'NM':
```

pmli reviewed (changes requested):
Ok, this should actually fix the CI.
